### ISSUE TITLE
Tiered jitting: Review/shrink memory usage for non-default code versions

### ISSUE THREAD

**Issue by noahfalk** [dotnet/coreclr#12611]
The current implementation of some of the data structures in CodeVersionManager are probably larger than they need to be. For example NativeCodeVersionNode stores both a native code pointer and a MethodDesc*, however it could store a union of both (other parts of the runtime know how to lookup code ptr -> MethodDesc*) to save 1 pointer per version. We should fix at least the low hanging fruit without getting carried away. Anything that would take more than a day or two to implement and test should wait until we can point to a specific performance scenario that is impacted to better justify and measure our work.

**Comment by davmason** [comment]
I looked at making the native code pointer and MethodDesc be a union, but it would add nontrivial complexity to the code. There is an m_flags field we could borrow a bit from to indicate whether it is a native code pointer or MethodDesc, but we would have to take a lock to update the field since the flag and the field would have to be updated atomically.

Since we're doing this as a precaution, we should wait until we know we need to add this complexity before adding locking on this codepath.

I took a quick look at the rest of the code, and nothing obvious jumped out at me as easy fixes.
